Mathlib/GroupTheory/OreLocalization/Cardinality.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2024 Jz Pan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jz Pan
+-/
+import Mathlib.GroupTheory.OreLocalization.Basic
+import Mathlib.SetTheory.Cardinal.Arithmetic
+
+/-!
+
+# Cardinality of Ore localizations
+
+This file contains some results on cardinality of Ore localizations.
+
+-/
+
+universe u v
+
+open Cardinal
+
+namespace OreLocalization
+
+variable {R : Type u} [Monoid R] (S : Submonoid R) [OreLocalization.OreSet S]
+  (X : Type v) [MulAction R X]
+
+@[to_additive]
+theorem oreDiv_one_surj_of_finite_left [Finite S] :
+    Function.Surjective (fun x ↦ x /ₒ (1 : ↥S) : X → OreLocalization S X) := by
+  refine OreLocalization.ind fun x s ↦ ?_
+  obtain ⟨i, j, hne, heq⟩ := Finite.exists_ne_map_eq_of_infinite (α := ℕ) (s ^ ·)
+  wlog hlt : j < i generalizing i j
+  · exact this j i hne.symm heq.symm (hne.lt_of_le (not_lt.1 hlt))
+  use s ^ (i - (j + 1)) • x
+  rw [oreDiv_eq_iff]
+  refine ⟨s ^ j, (s ^ (j + 1)).1, ?_, ?_⟩
+  · change s ^ j • x = s ^ (j + 1) • s ^ (i - (j + 1)) • x
+    rw [← mul_smul, ← pow_add, Nat.add_sub_cancel' hlt, heq]
+  · simp_rw [SubmonoidClass.coe_pow, OneMemClass.coe_one, mul_one, pow_succ]
+
+@[to_additive]
+theorem oreDiv_one_surj_of_finite_right [Finite X] :
+    Function.Surjective (fun x ↦ x /ₒ (1 : ↥S) : X → OreLocalization S X) := by
+  refine OreLocalization.ind fun x s ↦ ?_
+  obtain ⟨i, j, hne, heq⟩ := Finite.exists_ne_map_eq_of_infinite (α := ℕ) (s ^ · • x)
+  wlog hlt : j < i generalizing i j
+  · exact this j i hne.symm heq.symm (hne.lt_of_le (not_lt.1 hlt))
+  use s ^ (i - (j + 1)) • x
+  rw [oreDiv_eq_iff]
+  refine ⟨s ^ j, (s ^ (j + 1)).1, ?_, ?_⟩
+  · change s ^ j • x = s ^ (j + 1) • s ^ (i - (j + 1)) • x
+    rw [← mul_smul, ← pow_add, Nat.add_sub_cancel' hlt, heq]
+  · simp_rw [SubmonoidClass.coe_pow, OneMemClass.coe_one, mul_one, pow_succ]
+
+@[to_additive]
+theorem numeratorHom_surj_of_finite [Finite S] :
+    Function.Surjective (numeratorHom (S := S)) :=
+  oreDiv_one_surj_of_finite_left S R
+
+@[to_additive]
+theorem card_le_max : #(OreLocalization S X) ≤ max (lift.{v} #S) (lift.{u} #X) := by
```

Comment:
Elsewhere I think the convention is to call this `Cardinal.mk_oreLocalization_le`.
If we really want to use `card` for this, we should rename `Cardinal.mk` to `Cardinal.card`, which is certainly out of scope for this PR.

Reply:
Since it's the property of `OreLocalization`, I don't think it's a good idea to put it into namespace `Cardinal`.

The `IsLocalization.card_le` and `IsLocalization.card` is the existing result in mathlib which already uses this naming.
But perhaps you're right, the other results use different naming, for example [Polynomial.cardinal_mk_eq_max](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Polynomial/Cardinal.html#Polynomial.cardinal_mk_eq_max).
Perhaps we should call `OreLocalization.cardinal_mk_le_max`, etc.? But I'd argue that `card` is an abbreviation of `cardinal_mk`.

Reply:
Yeah, I find the comb `Finset.card`, `Fintype.card`, `IsLocalization.card` pretty poor. `IsLocalization.card_eq` could work, but really the function is called `Cardinal.mk` so `cardinalMk` is what you should use

Reply:
> `Finset.card` and `Fintype.card` are already functions, so I think they don't matter

I meant that their existence means that `IsLocalization.card` sounds like it should be a function

Reply:
Is current naming (`cardinal_mk_XXX`) OK?

Reply:
It's "ok" but the convention-approved name is `cardinalMk` 😄

Reply:
> It's "ok" but the convention-approved name is `cardinalMk` 😄

Currently mathlib has no results named `cardinalMk` yet; all of them uses `cardinal_mk`. So I think it's OK to use `cardinal_mk` in this PR. Perhaps rename all of them to `cardinalMk` is for another PR.

Reply:
I've opened #18808

Reply:
Done.